src/base/base.c

```diff
@@ -874,18 +885,57 @@ static bool should_run_user_handler_wrapper(zend_execute_data *execute_data)
 #endif
 }
 
+static zend_extension *xdebug_find_xdebug_extension()
+{
+	zend_llist_element *ext_ptr = zend_extensions.head;
+
+	do
+	{
+		zend_extension *zext = (zend_extension *)ext_ptr->data;
+
+		if (strcmp(zext->name, "Xdebug") == 0) {
+			return zext;
+		}
+
+		ext_ptr = ext_ptr->next;
+	} while (ext_ptr != NULL);
+
+	return NULL;
+}
+
+static void disable_statement_extension_handler()
+{
+	if (xdebug_extension == NULL) {
+		xdebug_extension = xdebug_find_xdebug_extension();
+	}
+	if (xdebug_extension != NULL) {
+		existing_statement_handler = xdebug_extension->statement_handler;
+		xdebug_extension->statement_handler = NULL;
+	}
+}
+
+static void enable_statement_extension_handler()
+{
+	if (xdebug_extension == NULL) {
+		xdebug_extension = xdebug_find_xdebug_extension();
+	}
+	if (xdebug_extension != NULL) {
+		xdebug_extension->statement_handler = existing_statement_handler;
+	}
+}
+
 /* We still need this to do "include", "require", and "eval" */
 static void xdebug_execute_ex(zend_execute_data *execute_data)
 {
 	bool run_user_handler = should_run_user_handler_wrapper(execute_data);
 
-	if (run_user_handler) {
+	if (run_user_handler && (XG_DBG(debugger_disabled) == 0 || execute_data->func->type == ZEND_EVAL_CODE)) {
```

Comment:
We only call the `xdebug_execute_user_code_begin()` and `xdebug_execute_user_code_end()`  functions if the debugger has not been disabled. Notice that we cannot just use a single if statement for both functions because the status of the debugger_disabled flag can change during the call to the `xdebug_old_execute_ex()` function

We also need to call the begin function if we are entering into some eval code because we want to record this eval code so that it can be used later